PrestaShop/PrestaShop
# Feature flags improvements

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This is an improvement of feature flags => See the description below or related issue.
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | See the description below. Running campaign https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/5485295067
| Fixed ticket?     | Fixes #32698 
| Related PRs       | 
| Sponsor company   | 

## What I have done?
- First of all, I have add DotEnv support for admin and console entry point.
- Now, we have some Handlers class aim to use Feature Flags: 
    - Database Layer
    - Env Layer
    - Query Layer
    - DotEnv Layer
- On top of that, we have a FeatureFlagManager aim to instanciate the used handler by the feature flags, in function of priority set in `type` column. exemple: If a feature flag has `env,query,dotenv,db` in type column, this class will check if it can be possible to get a value from environnement, if not, it check if we can retrieve value by query, if not it check if we can use dotenv file, and if not, we use DB (that can be always used!)
- I added `type` display in `New & experimental features` page, and the handler used is in bold for each feature flags.
- I added `type` display in `prestashop:feature-flags list` command, and the handler used is surrounded by brackets.
- I added a interface called `FeatureFlagStateCheckerInterface` to retrieve feature flag states simply.

## Database Layer (type: `db`)
- It's the same behaviour than before!

## DotEnv Layer (type `dotenv`)
- We use `.env` files in function of current environnement (dev, prod, etc...).
- Feature flags can be edited as well in the `New & experimental features` page, and the Handler will localise the `.env` file used to set this feature flag.
- The variable name in dotenv files must be: `PS_FF_{UPPERCASE FEATURE FLAG NAME}`.

## Env Layer (type `env`)
- **Cannot be editable!**
- The variable name in env must be: `PS_FF_{UPPERCASE FEATURE FLAG NAME}`.

## Query Layer (type `query`)
- **Cannot be editable and only in ModeDev!**
- The variable name in query string must be: `PS_FF_{UPPERCASE FEATURE FLAG NAME}`.

## How to test?
- By default, on a fresh install, we pass all feature flags in `db` type only.
- But, you can edit `PREFIX_feature_flag` table to add `env,query,dotenv,db` on the one you want.
- If you want to test Env Layer =>
    - In vhost of your installation, you can add: `SetEnv PS_FF_{FEATURE_FLAG_NAME} true`.
    - See in `New & experimental features` page the value of this feature flag and the handler used!
- If you want to test Query Layer =>
    - Add in your query string: `PS_FF_{FEATURE_FLAG_NAME}=true`.
    - See in `New & experimental features` page the value of this feature flag and the handler used!
- If you want to test DotEnv Layer =>
    - You can create a `.env` file in project root, and add `PS_FF_{FEATURE_FLAG_NAME}=true` in it.
    - See in `New & experimental features` page the value of this feature flag and the handler used!
    - Disable this feature flag from admin page, and check if your `.env` file has changed.
